app/gui/view/graph-editor/src/component/edge.rs

```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
+            split_arc: default(),
+            hover_sections: default(),
+            corner_points: default(),
+            target_attachment: default(),
+            hover_position: default(),
+            disabled: default(),
+            previous_target: default(),
+            previous_is_hoverable: default(),
+            previous_hover_split: default(),
+            previous_color: default(),
+            previous_target_attached: default(),
         }
+    }
 
-        impl $name {
-            /// Constructor.
-            #[allow(clippy::vec_init_then_push)]
-            pub fn new() -> Self {
-                let display_object = display::object::Instance::new_named(stringify!($name));
-                $(let $field = <$field_type>::new();)*
-                $(display_object.add_child(&$field);)*
-                let mut shape_view_events:Vec<PointerTarget_DEPRECATED> = Vec::default();
-                $(shape_view_events.push($field.events_deprecated.clone_ref());)*
-                let shape_view_events = Rc::new(shape_view_events);
-
-                let mut shape_type_map:HashMap<display::object::Id,ShapeRole> = default();
-                $(shape_type_map.insert(EdgeShape::id(&$field), $field_shape_type);)*
-                let shape_type_map = Rc::new(shape_type_map);
-
-                Self {display_object,shape_view_events,shape_type_map,$($field),*}
-            }
-
-            fn get_shape(&self, id:display::object::Id) -> Option<&dyn EdgeShape> {
-                match id {
-                    $(id if id == EdgeShape::id(&self.$field) => Some(&self.$field),)*
-                    _ => None,
-                }
-            }
+    /// Set the color of the edge.
+    fn set_color_and_redraw(&self, color: color::Lcha) {
+        let color_rgba = color::Rgba::from(color);
+        self.color.set(color_rgba);
+        self.redraw();
+    }
 
-            fn get_shape_type(&self, id:display::object::Id) -> Option<ShapeRole> {
-                self.shape_type_map.get(&id).cloned()
-            }
-        }
+    fn set_source_width(&self, width: f32) {
+        self.source_width.set(width);
+    }
 
-        impl Default for $name {
-            fn default() -> Self {
-                Self::new()
-            }
-        }
+    fn set_source_height(&self, height: f32) {
+        self.source_height.set(height);
+    }
 
-        impl display::Object for $name {
-            fn display_object(&self) -> &display::object::Instance {
-                &self.display_object
-            }
-        }
+    fn set_disabled(&self, disabled: bool) {
+        self.disabled.set(disabled);
+    }
 
-        impl AnyEdgeShape for $name {
-            #[allow(clippy::vec_init_then_push)]
-            fn shapes(&self) -> Vec<&dyn EdgeShape> {
-                let mut output = Vec::<&dyn EdgeShape>::default();
-                $(output.push(&self.$field);)*
-                output
-            }
+    fn source_x_range(&self) -> f32 {
```

Comment:
What is `source_x_range`? I'd expect `Range<f32>` as the return type for anything with `range` in the name.